File: streak-backend/src/api/jobs/job.model.ts
import mongoose, { Document, Schema } from 'mongoose';

/**
 * Enumeraciones para el modelo Job
 */
export enum JobStatus {
  DRAFT = 'draft',
  OPEN = 'open',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
}

export enum JobType {
  FIXED_PRICE = 'fixed_price',
  HOURLY = 'hourly',
}

/**
 * Interface para el documento Job
 */
export interface IJob extends Document {
  // Información básica
  title: string;
  description: string;
  category: string;
  subcategory?: string;

  // Cliente
  clientId: mongoose.Types.ObjectId;

  // Tipo y presupuesto
  type: JobType;
  budget?: number; // Para fixed_price
  hourlyRate?: {
    min: number;
    max: number;
  }; // Para hourly

  // Detalles del proyecto
  skillsRequired: string[];
  experienceLevel: 'entry' | 'intermediate' | 'expert';
  estimatedDuration?: string;
  attachments?: string[];

  // Estado
  status: JobStatus;
  isUrgent: boolean;

  // Freelancer asignado (cuando está in_progress)
  assignedFreelancerId?: mongoose.Types.ObjectId;

  // Estadísticas
  stats: {
    views: number;
    applicationsCount: number;
    savedCount: number;
  };

  // Fechas
  deadline?: Date;
  startDate?: Date;
  completionDate?: Date;
  createdAt: Date;
  updatedAt: Date;
}

/**
 * Schema de Job
 */
const JobSchema = new Schema<IJob>(
  {
    title: {
      type: String,
      required: [true, 'El título es requerido'],
      trim: true,
      minlength: [10, 'El título debe tener al menos 10 caracteres'],
      maxlength: [100, 'El título no puede exceder 100 caracteres'],
    },
    description: {
      type: String,
      required: [true, 'La descripción es requerida'],
      minlength: [50, 'La descripción debe tener al menos 50 caracteres'],
      maxlength: [5000, 'La descripción no puede exceder 5000 caracteres'],
    },
    category: {
      type: String,
      required: [true, 'La categoría es requerida'],
    },
    subcategory: String,

    clientId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },

    type: {
      type: String,
      enum: Object.values(JobType),
      required: true,
    },
    budget: {
      type: Number,
      min: [0, 'El presupuesto no puede ser negativo'],
    },
    hourlyRate: {
      min: {
        type: Number,
        min: [0, 'La tarifa mínima no puede ser negativa'],
      },
      max: {
        type: Number,
        min: [0, 'La tarifa máxima no puede ser negativa'],
      },
    },

    skillsRequired: {
      type: [String],
      validate: {
        validator: (v: string[]) => v && v.length > 0,
        message: 'Se requiere al menos una habilidad',
      },
    },
    experienceLevel: {
      type: String,
      enum: ['entry', 'intermediate', 'expert'],
      default: 'intermediate',
    },
    estimatedDuration: String,
    attachments: [String],

    status: {
      type: String,
      enum: Object.values(JobStatus),
      default: JobStatus.DRAFT,
    },
    isUrgent: {
      type: Boolean,
      default: false,
    },

    assignedFreelancerId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },

    stats: {
      views: { type: Number, default: 0, min: 0 },
      applicationsCount: { type: Number, default: 0, min: 0 },
      savedCount: { type: Number, default: 0, min: 0 },
    },

    deadline: Date,
    startDate: Date,
    completionDate: Date,
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

/**
 * Índices
 */
JobSchema.index({ clientId: 1, status: 1 });
JobSchema.index({ status: 1, createdAt: -1 });
JobSchema.index({ category: 1, status: 1 });
JobSchema.index({ skillsRequired: 1 });
JobSchema.index({ 'stats.applicationsCount': -1 });
JobSchema.index({ createdAt: -1 });

/**
 * Virtual: Cliente
 */
JobSchema.virtual('client', {
  ref: 'User',
  localField: 'clientId',
  foreignField: '_id',
  justOne: true,
});

/**
 * Virtual: Freelancer asignado
 */
JobSchema.virtual('assignedFreelancer', {
  ref: 'User',
  localField: 'assignedFreelancerId',
  foreignField: '_id',
  justOne: true,
});

/**
 * Exportar modelo
 */
export const Job = mongoose.model<IJob>('Job', JobSchema);

File: streak-app/lib/data/models/job_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'job_model.freezed.dart';
part 'job_model.g.dart';

/// Modelo de Trabajo/Job - Sincronizado con el backend
@freezed
class JobModel with _$JobModel {
  const factory JobModel({
    @JsonKey(name: '_id') required String id,
    required String title,
    required String description,
    required String category,
    String? subcategory,
    @JsonKey(name: 'clientId') ClientModel? clientId, // Cambiado de String a ClientModel?
    required String type, // 'fixed_price' | 'hourly'
    double? budget, // Para fixed_price
    HourlyRateModel? hourlyRate, // Para hourly
    @Default([]) @JsonKey(name: 'skillsRequired') List<String> skillsRequired,
    @Default('intermediate') String experienceLevel, // 'entry' | 'intermediate' | 'expert'
    String? estimatedDuration,
    @Default([]) List<String> attachments,
    required String status, // 'draft' | 'open' | 'in_progress' | 'completed' | 'cancelled'
    @Default(false) bool isUrgent,
    @JsonKey(name: 'assignedFreelancerId') String? assignedFreelancerId,
    required StatsModel stats,
    DateTime? deadline,
    DateTime? startDate,
    DateTime? completionDate,
    required DateTime createdAt,
    required DateTime updatedAt,
    // Campos populados
    ClientModel? client,
    UserModel? assignedFreelancer,
  }) = _JobModel;

  factory JobModel.fromJson(Map<String, dynamic> json) =>
      _$JobModelFromJson(json);
}

/// Modelo de Tarifa por Hora
@freezed
class HourlyRateModel with _$HourlyRateModel {
  const factory HourlyRateModel({
    required double min,
    required double max,
  }) = _HourlyRateModel;

  factory HourlyRateModel.fromJson(Map<String, dynamic> json) =>
      _$HourlyRateModelFromJson(json);
}

/// Modelo de Estadísticas
@freezed
class StatsModel with _$StatsModel {
  const factory StatsModel({
    @Default(0) int views,
    @Default(0) int applicationsCount,
    @Default(0) int savedCount,
  }) = _StatsModel;

  factory StatsModel.fromJson(Map<String, dynamic> json) =>
      _$StatsModelFromJson(json);
}

/// Modelo de Cliente (simplificado)
@freezed
class ClientModel with _$ClientModel {
  const factory ClientModel({
    @JsonKey(name: '_id') required String id,
    required String email,
    required ProfileModel profile,
  }) = _ClientModel;

  factory ClientModel.fromJson(Map<String, dynamic> json) =>
      _$ClientModelFromJson(json);
}

/// Modelo de Perfil
@freezed
class ProfileModel with _$ProfileModel {
  const factory ProfileModel({
    required String firstName,
    required String lastName,
    String? avatar,
  }) = _ProfileModel;

  factory ProfileModel.fromJson(Map<String, dynamic> json) =>
      _$ProfileModelFromJson(json);
}

/// Modelo de Usuario (simplificado)
@freezed
class UserModel with _$UserModel {
  const factory UserModel({
    @JsonKey(name: '_id') required String id,
    required String email,
    required ProfileModel profile,
  }) = _UserModel;

  factory UserModel.fromJson(Map<String, dynamic> json) =>
      _$UserModelFromJson(json);
}

/// Extensión para facilitar el uso del modelo
extension JobModelX on JobModel {
  bool get isOpen => status == 'open';
  bool get isDraft => status == 'draft';
  bool get isInProgress => status == 'in_progress';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
  
  bool get isFixedPrice => type == 'fixed_price';
  bool get isHourly => type == 'hourly';
  
  bool get hasDeadline => deadline != null;
  bool get isDeadlinePassed => deadline != null && deadline!.isBefore(DateTime.now());
  
  String get salaryRange {
    if (isFixedPrice && budget != null) {
      return '\$${budget!.toStringAsFixed(0)} (Proyecto fijo)';
    } else if (isHourly && hourlyRate != null) {
      return '\$${hourlyRate!.min.toStringAsFixed(0)} - \$${hourlyRate!.max.toStringAsFixed(0)} / hora';
    }
    return 'No especificado';
  }
  
  String get clientName {
    if (client != null) {
      return '${client!.profile.firstName} ${client!.profile.lastName}';
    }
    return 'Cliente';
  }
  
  String get experienceLevelText {
    switch (experienceLevel) {
      case 'entry':
        return 'Principiante';
      case 'intermediate':
        return 'Intermedio';
      case 'expert':
        return 'Experto';
      default:
        return 'Intermedio';
    }
  }
}

